src/main/java/com/flowingcode/vaadin/addons/gridexporter/BaseInputStreamFactory.java

```diff
@@ -60,14 +63,36 @@ protected Stream<T> getDataStream(Query newQuery) {
     }
     return stream;
   }
-  
-  protected List<Pair<String,Column<T>>> getGridHeaders(Grid<T> grid) {
-    return exporter.columns.stream().map(column -> ImmutablePair.of(GridHelper.getHeader(grid,column),column))
+
+  protected List<Pair<String, Column<T>>> getGridHeaders(Grid<T> grid) {
+    return exporter.columns.stream().map(column -> buildPair(grid, column))
         .collect(Collectors.toList());
   }
 
-  protected List<Pair<String,Column<T>>> getGridFooters(Grid<T> grid) {
-    return exporter.columns.stream().map(column -> ImmutablePair.of(GridHelper.getFooter(grid,column),column))
+  private Pair<String, Column<T>> buildPair(Grid<T> grid, Column<T> column) {
+    String header = (String) ComponentUtil.getData(column, GridExporter.COLUMN_HEADER);
+    if (Strings.isBlank(header)) {
+      header = GridHelper.getHeader(grid, column);
+    }
+    if (Strings.isBlank(header)) {
+      try {
+        Method getHeaderComponent = Column.class.getMethod("getHeaderComponent");
+        Element element = (Element) getHeaderComponent.invoke(column);
+        header = element.getTextRecursively();
+      } catch (NoSuchMethodException e) {
+        header = "";
+      } catch (IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
+        throw new IllegalStateException("Problem when trying to obtain the header component", e);
+      }
+    }
+
+    Pair<String, Column<T>> result = ImmutablePair.of(header, column);
+    return result;
+  }
+
+  protected List<Pair<String, Column<T>>> getGridFooters(Grid<T> grid) {
+    return exporter.columns.stream()
```

Comment:
It's not mentioned in #33, but the same approach can be implemented for footers.